### PR TITLE
Add API to listen to app state changes.

### PR DESCRIPTION
**Pull request by mandrigin**
An empty API to listen to the app status changes from status-react.

Part of #734

### CODE REVIEW

File: geth/api/app_state.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+const (
+	AppStateForeground = AppState("foreground")
+	AppStateBackground = AppState("background")
+	AppStateInactive   = AppState("inactive")
+
+	AppStateInvalid = AppState("")
+)
+
+func validAppStates() []AppState {
```
Comment (themue) [inline geth/api/app_state.go]: Why no constant here?

File: geth/api/app_state.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+	return []AppState{AppStateInactive, AppStateBackground, AppStateForeground}
+}
+
+func ParseAppState(stateString string) (AppState, error) {
```
Comment (themue) [inline geth/api/app_state.go]: Call me pedantic, but at least for all public functions, types, etc. I would like to see comments. Simply imagine the impression of people reading about our code at https://godoc.org/.

File: geth/api/backend.go [hunk 3]
```diff
@@ -236,3 +236,10 @@ func (b *StatusBackend) ConnectionChange(state ConnectionState) {
 	// logic of handling state changes here
 	// restart node? force peers reconnect? etc
 }
+
+// AppStateChange handles app state changes (background/foreground).
+func (b *StatusBackend) AppStateChange(state AppState) {
+	log.Info("App State changed: %s", state)
+
+	// TODO: put node in low-power mode
```
Comment (themue) [inline geth/api/backend.go]: Always, not depending on state? Also here the comment of the method could provide more information on what will happen based on the passed argument.

**Review by themue — CHANGES_REQUESTED** [review]
Idea mostly is clear but I'm one of the old documentation lovers helping later maintainers. 😁

File: geth/api/app_state.go (hunk 1 above)
Comment (mandrigin) [reply geth/api/app_state.go]: basically to make it _truly_ constant.
Arrays can only be declared with `var` and it is mutable.
No big deal though, I can change it.

File: geth/api/backend.go (hunk 3 above)
Comment (mandrigin) [reply geth/api/backend.go]: ah, yeah, I'll fix the comment

File: geth/api/app_state.go (hunk 1 above)
Comment (themue) [reply geth/api/app_state.go]: Expected it and yeah, sadly no real constant slices. In global cases I would do it similar.

File: geth/api/app_state.go (hunk 2 above)
Comment (mandrigin) [reply geth/api/app_state.go]: Hmm, I thought I fixed that. Maybe outdated diff?

File: geth/api/app_state.go (hunk 1 above)
Comment (mandrigin) [reply geth/api/app_state.go]: So, would you mind me keeping it as is or do you want me to make it a `var`?

File: geth/api/backend.go (hunk 3 above)
Comment (mandrigin) [reply geth/api/backend.go]: fixed

File: geth/api/app_state.go (hunk 1 above)
Comment (themue) [reply geth/api/app_state.go]: Just keep it, it's more robust and elegant. Maybe a little comment. 

```
// validAppStates returns an immutable set of valid states.
```

Comment (mandrigin) [reply geth/api/app_state.go]: done!

**Review by themue — APPROVED** [review]
LGTM

File: geth/api/app_state.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+const (
+	AppStateForeground = AppState("foreground")
+	AppStateBackground = AppState("background")
+	AppStateInactive   = AppState("inactive")
+
+	AppStateInvalid = AppState("")
+)
+
+// validAppStates returns an immutable set of valid states.
+func validAppStates() []AppState {
```
Comment (adambabik) [inline geth/api/app_state.go:26]: Maybe declare it as a `var`?

**Review by adambabik — COMMENTED** [review]
Not really sure we need to merge it as it does nothing for now. Do we have a plan how it can be useful? Maybe we can keep it in a long-running branch for now?

**Comment by mandrigin** [comment]
I'm not a huge fan of long branches of any kind.
Sooner than expected the cost of keeping it up-to-date skyrockets.
This one I want to merge and to connect to status-react ASAP, so then it would be possible to introduce this "low-power" mode of a node.

That's an important part of energy efficiency optimization, that is one more relatively simple fix (to that make the app eat less battery in background).

IDK what the exact optimization could be there, I have a couple of options, but I want to try them in a real app.

File: geth/api/app_state.go (hunk 4 above)
Comment (mandrigin) [reply geth/api/app_state.go:26]: there is [a discussion about it](https://github.com/status-im/status-go/pull/735#discussion_r174515431) in this review already.
in general, I don't like that `var` is mutable.

**Comment by mandrigin** [comment]
In general, I much prefer doing multiple small PRs when implementing a project.
They are easy to review and have very limited regression scope.

**Comment by mandrigin** [comment]
@adambabik but yeah, you reminded me that I didn't really publish the next steps for what can be optimized. Thanks! Now I did :) https://github.com/status-im/ideas/issues/83#issuecomment-373103508

**Comment by adambabik** [comment]
> I'm not a huge fan of long branches of any kind.
Sooner than expected the cost of keeping it up-to-date skyrockets.

Well... This can be a long discussion I don't want to start :P 

My problem with such changes is that they kind of pollute the code base because they are not useful at the moment. They are kind of loose ends that wait for better times which may never arrive. If you strongly believe it will evolve into something concrete then let's merge it. 

My approach would be to create a branch for an idea and just keep it there. It's pretty separate from the other parts of the code so maintaining it should be easy.

**Comment by mandrigin** [comment]
@adambabik I see your point, that is Speculative Generality, when we introduce these unused methods.
Anyway, I'm planning to continue working on this idea this and next week, so it won't be unused for long :)

**Review by adambabik — APPROVED** [review]

**Pull request merged**